### PR TITLE
docs: get docs from checker comments

### PR DESCRIPTION
**Pull request by fexolm**

### CODE REVIEW

**Comment by fexolm** [comment]
Ready for appendCombine, please check

File: cmd/makedocs/main.go [hunk 1]
```diff
@@ -2,48 +2,73 @@ package main
...
+			continue
+		}
+		c := checker{
+			Name:            r.Name(),
+			SyntaxOnly:      r.SyntaxOnly,
+			Experimental:    r.Experimental,
+			VeryOpinionated: r.VeryOpinionated,
+		}
+		for _, comment := range f.Comments {
+			if strings.HasPrefix(comment.Text(), "!") {
```
Comment (Quasilyte) [inline cmd/makedocs/main.go]: `Text()` computes combined text every time you call it.
Maybe it's enough to check first comment from group test for that prefix?
Or re-use `Text()` result.
Just making sure you know what's going on here. There should be no real performance issue here.

File: cmd/makedocs/main.go [hunk 2]
```diff
@@ -59,15 +84,62 @@ func main() {
...
-	b, err := ioutil.ReadFile(checkersPath + name + ".md")
-	return string(b), err
-}
+type parseState uint
+
+const (
+	emptyLine parseState = 1 << iota
+	desciption
+	before
+	after
```
Comment (Quasilyte) [inline cmd/makedocs/main.go]: Usually, we prefix enum values with some common prefix that resembles their type.
`stEmptyLine`, `stDescriotion`, etc.
Yeah, like in C. :)

File: cmd/makedocs/main.go [hunk 3]
```diff
@@ -59,15 +84,62 @@ func main() {
...
+)
+
+func parseComment(text string, c *checker) {
+	lines := strings.Split(text, "\n")
+	s := emptyLine
```
Comment (Quasilyte) [inline cmd/makedocs/main.go]: I would prefer `s/s/st/` to avoid confusion with "string".
(Optional.)

File: cmd/makedocs/main.go (hunk 1 above)
Comment (cristaloleg) [reply cmd/makedocs/main.go]: From the other side: how many comments it will process? Hundreds at most ;)

Comment (Quasilyte) [reply cmd/makedocs/main.go]: Yep, this is why it doesn't matter.
Just making a note that `Text()` is not a simple getter.

**Comment by Quasilyte** [comment]
What about more explicit state machine?
(pseudo-go):
```go
grammar := []syntax{
  stx.Summart,
  stx.Or(stx.All(stx.EmptyLine, stx.Details, stx.EmptyLine), stx.EmptyLine),
  stx.Before,
  stx.EmptyLine,
  stx.After,
  stx.Or(stx.All(stx.EmptyLine, stx.Notice, stx.End), stx.End),
}
```
Each phase captures state and can fill doc being generated.
It's also possible to print which exact phase failed, during which alternation, which element was expected, etc.

(This is only an idea, not request for re-implementation.)

**Comment by fexolm** [comment]
@Quasilyte @cristaloleg please watch

File: cmd/makedocs/main.go [hunk 4]
```diff
@@ -59,15 +86,82 @@ func main() {
...
+	return nil
 }
 
-func shortDescription(desc string) string {
-	// We expect every checker to have first doc line
-	// to be a short summary, complete sentence.
-	summary := strings.Split(desc, "\n")[0]
-	summary = strings.TrimSpace(summary)
-	return summary
+func parseDesc(lines []string, ind *int, c *checker) error {
```
Comment (cristaloleg) [inline cmd/makedocs/main.go]: s/ind/index/g 😉

File: cmd/makedocs/main.go [hunk 5]
```diff
@@ -12,38 +17,60 @@ import (
...
+		if !ok {
+
+			log.Printf("File not found: %s%s_checker.go", checkersPath, r.Name())
+			continue
+		}
+		c := checker{
+			Name:            r.Name(),
+			SyntaxOnly:      r.SyntaxOnly,
+			Experimental:    r.Experimental,
+			VeryOpinionated: r.VeryOpinionated,
```
Comment (Quasilyte) [inline cmd/makedocs/main.go]: You could store `*lint.Rule` inside `checker` and do just: `c := checker{rule: rule}`.
Current approach also makes sense, so this is not a request for changes.

File: cmd/makedocs/main.go [hunk 6]
```diff
@@ -59,15 +86,82 @@ func main() {
...
+		return errors.New("parseDesc: no description provided")
+	}
+	if strings.HasPrefix(lines[*index], "@") { // if no description
+		return nil
+	}
+	for *index < len(lines) && len(lines[*index]) > 0 {
+		c.Description += strings.TrimSpace(lines[*index]) + "\n"
+		*index++
+	}
+	*index++ //skip empty line
```
Comment (Quasilyte) [inline cmd/makedocs/main.go]: You could do every stage parsing function a method of some type, which would store index inside itself,
and do something like:
```go
p.skipEmptyLine()
```
Which would be more readable and can also perform a check that current like is really empty or not.

Comment (Quasilyte) [reply cmd/makedocs/main.go]: Can be refactored later.

File: cmd/makedocs/main.go [hunk 7]
```diff
@@ -59,15 +86,82 @@ func main() {
...
+func parseBefore(lines []string, index *int, c *checker) error {
+	if len(lines) <= *index || strings.TrimSpace(lines[*index]) != "@Before:" {
+		return errors.New("parseBefore: no @Before: section found")
+	}
+	*index++
+	for *index < len(lines) && len(lines[*index]) > 0 {
+		c.Before += strings.TrimSpace(lines[*index]) + "\n"
+		*index++
+	}
+	*index++ //skip empty line
```
Comment (Quasilyte) [inline cmd/makedocs/main.go]: Nit: no space between `//` and comment text.
```diff
- //skip empty line
+ // skip empty line
```
I can fix that for you.

**Comment by Quasilyte** [comment]
Also, thank you for doing this. :)

**Pull request merged**